### PR TITLE
Initial BuildKite implementation

### PR DESCRIPTION
**Pull request by KevinGrandon**
This adds a BuildKite pipeline which builds two Docker images and runs
multiple tasks in parallel.

A couple things to note:

- Unit tests are sharded across multiple machines.
- AWS instances are distributed via an AWS Elastic Container Registry
  that we own.

The pipeline looks like this:
Pipeline Screenshot

Pros:

- Great parallelism and pipeline controls.
- Lots of knobs to turn to improve performance, and get down to ~1m
  builds.

Cons:

- Currently right now you need a buildkite account to see build pages.
  This is something that Buildkite is working on releasing, and if I had
  to guess would be ready in the next few months.

Steps after landing:

- Enable the BuildKite webhook, and update the BuildKite project to be
  this repository.
- Run BuildKite and travis in parallel until Buildkite is in good
  enough shape to replace Travis.
- Update Codecov token for pipeline.

### CODE REVIEW

**Comment by peter-edge** [comment]
This is a great start! Everyone notice the total build time of ~4m30s, with room for making it faster.

Will probably want to pull back some of the docker-compose stuff from https://github.com/yarpc/yarpc-go/pull/1368 to make an equivalent CI command for local testing.

**Comment by abhinav** [comment]
This is exciting! I'm all for it.

CC @breerly @kriskowal @willhug

(There's a related email in your inboxes too.)

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        image-repository: 296822479253.dkr.ecr.us-east-2.amazonaws.com/fusionjs
+  - name: ":docker: :package: 1.9"
+    plugins:
+      docker-compose#v1.5.2:
+        build: yarpc-go-1.9
+        image-repository: 296822479253.dkr.ecr.us-east-2.amazonaws.com/fusionjs
+  - wait
+  - name: ":go: 1.8 test - %n"
+    command: "make test"
+    parallelism: 2
```
Comment (KevinGrandon) [inline .buildkite/pipeline.yml]: It's easy to tune this number as needed to change the parallel worker count.

File: internal/shard/main.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"strconv"
+)
+
+var usage = fmt.Sprintf("Usage: %s shardNum totalShards args...", os.Args[0])
```
Comment (abhinav) [inline internal/shard/main.go]: what exactly does this executable do? can we add some docs?

File: etc/make/local.mk [hunk 3]
```diff
@@ -126,12 +126,12 @@ basiclint: gofmt govet golint staticcheck errcheck # run gofmt govet golint stat
...
-test: $(THRIFTRW) __eval_packages ## run all tests
-	PATH=$(BIN):$$PATH go test -race $(PACKAGES)
+test: $(THRIFTRW) __eval_chunked_packages ## run chunked tests
+	PATH=$(BIN):$$PATH go test -race $(CHUNKED_PACKAGES)
 
 .PHONY: cover
-cover: $(THRIFTRW) $(GOCOVMERGE) $(PARALLEL_EXEC) $(COVER) __eval_packages ## run all tests and output code coverage
-	PATH=$(BIN):$$PATH ./etc/bin/cover.sh $(PACKAGES)
+cover: $(THRIFTRW) $(GOCOVMERGE) $(PARALLEL_EXEC) $(COVER) __eval_chunked_packages ## run chunked tests and output code coverage
+	PATH=$(BIN):$$PATH ./etc/bin/cover.sh $(CHUNKED_PACKAGES)
```
Comment (abhinav) [inline etc/make/local.mk]: Won't this basically stomp on coverage from different shards until the last one is just the tests from that shard?

File: internal/shard/main.go (hunk 2 above)
Comment (KevinGrandon) [reply internal/shard/main.go]: This basically just splits a list of files into different chunks so we can run them in parallel. E.g., given four files, "a, b, c, and d", chunking them might return "a, c" and "b, d". Adding a comment with more information and usage might be good.

File: etc/make/local.mk (hunk 3 above)
Comment (KevinGrandon) [reply etc/make/local.mk]: From the codecov docs: https://docs.codecov.io/v4.3.6/docs/merging-reports
> Codecov does not override report data for multiple uploads. We always merge the data. Simply upload all three reports at once, or separately.

So this should just merge the parallel codecov runs. I'll do some more verification on my branch, but this should just work.

**Comment by peter-edge** [comment]
Hey all, should we close this until Buildkite has public build pages? Do we want to push forward on it?

**Comment by KevinGrandon** [comment]
> Hey all, should we close this until Buildkite has public build pages? Do we want to push forward on it?

Up to you all. I'll push on getting a window for public build pages. I'm happy to help pick this up again after that.

My recommendation here is actually to run both buildkite and travis in parallel for a while, and not block PRs on buildkite failures. That way we can:

* Use Buildkite and evaluate alongside Travis.
* Try out the ergonomics of Buildkite, while still relying on Travis.
* Have the ability to invite any collaborators to a Buildkit yarpc team who can view build pages, and have faster CI response on PRs. (Should lead to faster developer velocity).

There was a concern around duplicate coverage, but since codecov would simply combine coverage reports there should be no difference here. E.g., whether it's calls=2 per line, or calls=4, that line would still be considered covered, and should not lead to false positivies.

**Comment by peter-edge** [comment]
Closing this for now, but keep the branch around - want to wait for public build pages

**Comment by abhinav** [comment]
Resurrecting this because of in-team discussions. @KevinGrandon I'll reach out to you internally.

**Pull request merged**